File: Monotonic/next-greater-element.js
/*The next greater element of some element x in an array is the first greater element that is to the right of x in the same array.

You are given two distinct 0-indexed integer arrays nums1 and nums2, where nums1 is a subset of nums2.

For each 0 <= i < nums1.length, find the index j such that nums1[i] == nums2[j] and determine the next greater element of nums2[j] in nums2. If there is no next greater element, then the answer for this query is -1.

Return an array ans of length nums1.length such that ans[i] is the next greater element as described above.*/

const nextGreaterElement = (nums1, nums2) => {
    const nextGreaterElementMap = new Map();
    const stack = [];

    for (let i = 0; i < nums2.length; i++) {
        while (stack.length > 0 && stack[stack.length - 1] < nums2[i]) {
            nextGreaterElementMap.set(stack.pop(), nums2[i]);
        }
        
        stack.push(nums2[i]);
    }

    for (let num of stack) {
        nextGreaterElementMap.set(num, -1);
    }

    const ans = [];
    for (let i = 0; i < nums1.length; i++) {
        ans[i] = nextGreaterElementMap.get(nums1[i]);
    }

    return ans;
}

const nums1 = [4,1,2], nums2 = [1,3,4,2];

console.log(nextGreaterElement(nums1, nums2));

File: Monotonic/next_greater_element.py
from typing import List


class NextGreaterElement:

    def next_greater_element(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nextGreaterElementsMap = {}
        monoDecreasingStack = []
        for i in range(len(nums2)):
            num = nums2[i]
            nextGreaterElementsMap[num] = -1
            while monoDecreasingStack and num > monoDecreasingStack[-1]:
                nextGreaterElementsMap[monoDecreasingStack[-1]] = num
                monoDecreasingStack.pop()
            monoDecreasingStack.append(num)
        
        nextGreaterElements = []
        for num in nums1:
            nextGreaterElements.append(nextGreaterElementsMap[num])
        
        return nextGreaterElements

nums1 = [4,1,2]
nums2 = [1,3,4,2]
print(NextGreaterElement().next_greater_element(nums1, nums2))